node/pkg/watchers/solana/ccq.go

```diff
@@ -0,0 +1,200 @@
+	select {
+	case w.queryResponseC <- queryResponse:
+		w.ccqLogger.Debug("published query response error to handler")
+	default:
+		w.ccqLogger.Error("failed to published query response error to handler")
+	}
+}
+
+// ccqHandleQuery is the top-level query handler. It breaks out the requests based on the type and calls the appropriate handler.
+func (w *SolanaWatcher) ccqHandleQuery(ctx context.Context, queryRequest *query.PerChainQueryInternal) {
+
+	// This can't happen unless there is a programming error - the caller
+	// is expected to send us only requests for our chainID.
+	if queryRequest.Request.ChainId != w.chainID {
+		panic("ccqevm: invalid chain ID")
+	}
+
+	start := time.Now()
+
+	switch req := queryRequest.Request.Query.(type) {
+	case *query.SolanaAccountQueryRequest:
+		w.ccqHandleSolanaAccountQueryRequest(ctx, queryRequest, req)
+	default:
+		w.ccqLogger.Warn("received unsupported request type",
+			zap.Uint8("payload", uint8(queryRequest.Request.Query.Type())),
+		)
+		w.ccqSendQueryResponse(queryRequest, query.QueryFatalError, nil)
+	}
+
+	query.TotalWatcherTime.WithLabelValues(w.chainID.String()).Observe(float64(time.Since(start).Milliseconds()))
+}
+
+// ccqHandleSolanaAccountQueryRequest is the query handler for a sol_account request.
+func (w *SolanaWatcher) ccqHandleSolanaAccountQueryRequest(ctx context.Context, queryRequest *query.PerChainQueryInternal, req *query.SolanaAccountQueryRequest) {
+	requestId := "sol_account:" + queryRequest.ID()
+	w.ccqLogger.Info("received a sol_account query", zap.String("requestId", requestId))
+
+	rCtx, cancel := context.WithTimeout(ctx, rpcTimeout)
+	defer cancel()
+
+	// Convert the accounts from byte arrays to public keys.
+	accounts := make([]solana.PublicKey, 0, len(req.Accounts))
+	for _, acct := range req.Accounts {
+		accounts = append(accounts, acct)
+	}
+
+	// Create the parameters needed for the account read and add any optional parameters.
+	params := rpc.GetMultipleAccountsOpts{
+		Encoding:   solana.EncodingBase64,
+		Commitment: rpc.CommitmentType(req.Commitment),
+	}
+
+	if req.MinContextSlot != 0 {
+		params.MinContextSlot = &req.MinContextSlot
+	}
+
+	if req.DataSliceLength != 0 {
+		params.DataSlice = &rpc.DataSlice{
+			Offset: &req.DataSliceOffset,
+			Length: &req.DataSliceLength,
+		}
+	}
+
+	// Read the accounts.
+	info, err := w.rpcClient.GetMultipleAccountsWithOpts(rCtx, accounts, &params)
+	if err != nil {
+		w.ccqLogger.Error("read failed for sol_account query request",
+			zap.String("requestId", requestId),
+			zap.Any("accounts", accounts),
+			zap.Any("params", params),
+			zap.Error(err),
+		)
+
+		w.ccqSendQueryResponse(queryRequest, query.QueryRetryNeeded, nil)
+		return
+	}
+
+	// Read the block for this slot to get the block time.
+	maxSupportedTransactionVersion := uint64(0)
+	block, err := w.rpcClient.GetBlockWithOpts(rCtx, info.Context.Slot, &rpc.GetBlockOpts{
```

Comment:
should these two requests be done in a batch? since the block is only referenced by slot in the return of `GetMultipleAccountsWithOpts`, I'm concerned these could be different between two http requests

Reply:
I don't see how these can be in a batch since I need the slot number returned by `GetMultipleAccountWithOpts` in order to call `GetBlockWithOpts`.

Reply:
ah 😦 I'll leave the convo open for other reviewers to be aware / think through

Reply:
I think this should be better with the restriction of `finalized`